File: c/cbuf/cbuf.c
// cbuf.c -- circular buffer
// variation on: https://en.wikipedia.org/wiki/Circular_buffer

#include <stdio.h>
#include <stdlib.h>

typedef struct _t_cbuf {
    double *buf; /* array as circular buffer of `size_t` ints */
    int end;     /* write index */
    int start;   /* read index */
    size_t size; /* size of buffer */
} t_cbuf;

t_cbuf *cbuf_init(size_t size);
void cbuf_put(t_cbuf *c, double item);
double cbuf_get(t_cbuf *c);
void cbuf_free(t_cbuf *c);

t_cbuf *cbuf_init(size_t size) {
    t_cbuf *cbuf = malloc(sizeof(t_cbuf));
    if (!cbuf || !size) {
        printf("error: !cbuf || !size");
        return NULL;
    }
    cbuf->end = 0;
    cbuf->start = 0;
    cbuf->size = size;
    cbuf->buf = (double *)calloc(size, sizeof(double));
    if (!cbuf->buf) {
        free(cbuf);
        return NULL;
    }
    return cbuf;
}

void cbuf_put(t_cbuf *c, double item) {
    c->buf[c->end++] = item;
    c->end %= c->size;
}

double cbuf_get(t_cbuf *c) {
    double item = c->buf[c->start++];
    c->start %= c->size;
    return item;
}

void cbuf_free(t_cbuf *c) {
    free(c->buf);
    free(c);
}

int main() {
    t_cbuf *c = cbuf_init(4);
    if (c) {
        double entry = 0.0;
        for (int i = 0; i < 10; i++) {
            entry += i;
            cbuf_put(c, entry);
            double res = cbuf_get(c);
            printf("%d: %f\n", i, res);
        }
        cbuf_free(c);
        return 0;
    }
    return -1;
}

File: lua/cbuf/cbuf.lua

-- cbuf: a circular buffer
-- see: inspired by wikipedia entry

BUFLEN = 4


function cbuf(size)
    local self = {
        buf = {},
        size = size,
        current = 0.0,
        end_ = 0,
        start = 0
    }

    local put = function(value)
        self.end_ = self.end_ + 1
        self.buf[self.end_] = value
        self.end_ = self.end_ % self.size
    end

    local get = function()
        self.start = self.start + 1
        self.current = self.buf[self.start]
        self.start = self.start % BUFLEN
        return self.current
    end

    local get_start = function()
        return self.start
    end

    local get_end = function()
        return self.end_
    end

    local dump = function()
        for i=1, self.size do
            print("buf[".. i .. "] = " .. self.buf[i])
        end
        print("start: " .. self.start)
        print("end: " .. self.end_)
        print("current: " .. self.current)
    end

    return {
        put = put,
        get = get,
        get_start = get_start,
        get_end = get_end,
        dump = dump
    }
end


function main()
    local c = cbuf(4)
    local entry = 0.0
    for i=1, 10 do
        entry = entry + i
        c.put(entry)
        current = c.get()
        -- print(i .. " current: " .. current)
        print(i .. " start: " .. c.get_start() .. " current: " .. current)
    end
    -- c.dump()
end
